multiqc/modules/atropos/atropos.py

```diff
@@ -0,0 +1,1356 @@
+        phase = mod_cust_config.get('phase', 'trim')
+        submodule_class = SUBMODULES[phase]
+        self.submodule = submodule_class()
+
+        # Initialise the parent object
+        super(MultiqcModule, self).__init__(
+            name=name or self.submodule.name,
+            anchor=anchor or self.submodule.anchor,
+            href=ATROPOS_GITHUB_URL,
+            info=info or self.submodule.info + (
+                "is a general-purpose NGS pre-processing tool that "
+                 "specializes in adatper- and quality-trimming, written "
+                 "by John Didion at NHGRI. It is a fork of Cutadapt, "
+                 "written by Marcel Martin at Stockholm University."))
+
+        self.css = {}
+        self.js = {}
+
+        # Load data from logs and generate the report
+        if from_config:
+            self.init_from_config()
+            self.submodule.atropos_report(self)
+
+    def __getattr__(self, name):
+        """Fetch all missing attributes from the submodule.
+        """
+        return getattr(self.submodule, name)
+
+    def init_from_config(self):
+        """Initialize module from JSON files discovered via MultiQC
+        configuration.
+        """
+        log_files = self.find_log_files('atropos', filehandles=True)
+        if not log_files:
+            # Atropos summary files are JSON files (with .json extension) and
+            # always have '"program": "Atropos"' as the first key-value pair.
+            # Note that this way is deprecated; currently it produces a warning
+            # but it may produce an error in the future.
+            patterns = dict(fn='*.json', contents='"program": "Atropos"')
+            log_files = self.find_log_files(patterns, filehandles=True)
+
+        for file_dict in log_files:
+            fileobj = file_dict['f']
+            data = json.load(fileobj)
+            # We always run the trim submodule because it adds to the general
+            # stats table.
+            if self.submodule.can_add_data(data):
+                self.add_atropos_data(data, fileobj, file_dict)
+
+        if self.num_samples == 0:
+            log.debug("Could not find any reports in %s", config.analysis_dir)
+            raise UserWarning
+        else:
+            log.info("Found %s reports", self.num_samples)
+
+    def add_atropos_data(self, data, data_source, file_dict):
+        """For each sample, Atropos generates a <sample>.json file. That
+        file has summary information (numbers of files processed, description of
+        the inputs (e.g. FASTA vs FASTQ, single vs paired-end), and a summary
+        of the analyses performed.
+
+        Args:
+            data: Atropos summary dict.
+            data_source: The file from which 'data' was loaded.
+        """
+        close = False
+        if data_source and isinstance(data_source, str):
+            data_source = open(data_source)
+            close = True
+        if data is None:
```

Comment:
I think I must be missing something here - when can `data` be `None`?

Reply:
this is legacy code - I got rid of it